aptos-move/block-executor/src/executor.rs

```diff
@@ -1451,12 +1533,20 @@ where
                 executor_arguments,
                 signature_verified_block,
                 base_view,
+                false,
             )
         };
 
+        let resource_group_bcs_fallback = matches!(
+            ret,
+            Err(BlockExecutionError::FallbackToSequential(PanicOr::Or(
+                IntentionalFallbackToSequential::ResourceGroupSerializationError(_)
+            )))
+        );
+
         // Sequential execution fallback
         // Only worth doing if we did parallel before, i.e. if we did a different pass.
-        if self.config.local.concurrency_level > 1 {
+        if self.config.local.concurrency_level > 1 || resource_group_bcs_fallback {
```

Comment:
Suppose concurrency_level = 1 and the sequential execution returned ResourceGroupSerializationError, we still execute sequential execution with resource_group_bcs_fallback flag set to true?

Reply:
yes

Reply:
this is not good, because we could first see module fallback, and then when executing here see bcs fallback.

you need to leave the first fallback as it is (potentially skip it if error is ResourceGroupSerializationError), and then add a second fallback just for ResourceGroupSerializationError.

Reply:
yes, this does slow the module RW fallback, thought it may be worth it to not have a fallback to fallback. Without changing the return logic, then if the concurrency level was >1, we would do normal sequential, and if that had resource group error, we do second fallback, is what you also had in mind?

Reply:
this is not correct.

you would either need to pass "true" bellow always (i.e. fallback always does it), or have fallback to fallback (what I suggested)

currently if your transactions are :

`txn1, module_publish, txn2, txn_with_bcs_issue`

you will still panic

Reply:
no because parallel execution always returns bcs serialization error, i.e. this:
https://github.com/aptos-labs/aptos-core/blob/6020c61871b0925fae4ecd0a03e56c93fdc77f28/aptos-move/block-executor/src/executor.rs#L1125
right?
so if we go through parallel, we are guaranteed to run with true,
if we go through sequential, we will only do fallback if we have a bcs issue, and then it will be again true.

maybe too convoluted though. should I just explicitly pass 'true', or double fallback? i'm fine with either.

Reply:
oh that is hella confusing. 

can we return the right error, and have fallback just always do "true" for resource_group_bcs_fallback ?

Reply:
we don't know what the right error is, because everything short circuits and some aren't even possible to really order. Ideally, what I want to do is to return all errors (for tests), but I hadn't done that yet. For now the thinking was that any error triggers the fallback, and we can choose to configure it with serialization handling or not. From sequential it's much cleaner, everything makes sense.

Reply:
seems like we should then have IntentionalFallbackToSequential::FallbackFromParallelError, which always sets bcs to true.

it is going to be very confusing to return ResourceGroupSerializationError when that was never thrown at all

Reply:
Sounds good! will do that. And let me hardcode true there for even more clarity.

Reply:
should be done